File: Top Interview 150/01. Array and String/1. Merge Sorted Array/Java/Merge Sorted Array/src/MergeSortedArray.java
public class MergeSortedArray {

    public void merge(int[] nums1, int m, int[] nums2, int n) {

        //keeps track of which elements we are comparing
        int nums1Idx = m - 1;
        int nums2Idx = nums2.length - 1;
        int currIdx = nums1.length - 1;

        //helps with sorting nums1 array
        while(currIdx >= 0) {

            System.out.println("currIdx: " + currIdx);
            System.out.print("Before: ");
            printArray(nums1);

            if(nums2Idx >= 0 && nums2[nums2Idx] >= nums1[nums1Idx]) {
                nums1[currIdx] = nums2[nums2Idx];
                --nums2Idx;
                --currIdx;
            } else {
                nums1[currIdx] = nums1[nums1Idx];
                --nums1Idx;
                --currIdx;
            }

            System.out.print("After: ");
            printArray(nums1);
            System.out.println("");

        }

        System.out.println("End of Merge Function.");
    }

    private void printArray(int[] array) {

        System.out.print("[ ");

        for(int i = 0; i < array.length; ++i) {
            System.out.print(array[i] + " ");
        }

        System.out.println("]");
    }

}




//            //compare values in the nums2 array
//            if(nums2Idx >= 0 && nums2[nums2Idx] == 0) {
//                --nums2Idx;
//            } else if(nums2Idx >= 0) {
//
//            }
//
//            //compare values in the nums1 array
//            if(nums1Idx >= 0 && nums1[nums1Idx] == 0) {
//                --nums1Idx;
//            } else if(nums1Idx >= 0) {
//
//            }

File: Top Interview 150/01. Array and String/01. Merge Sorted Array/C++/v2/MergeSortedArray/MergeSortedArray/MergeSortedArray.cpp
#include <iostream>
#include "MergeSortedArray.h"
using namespace std;

void Solution::merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
	
	// Ensures the following constraints are met:
	// * nums1.length == m + n
	// * nums2.length == n
	// * nums2.empty == false
	if (nums1.size() == (m + n) && !nums2.empty() && nums2.size() == n) {
	
		//keeps track of the current element you're comparing
		int i = nums1.size() - 1;

		//Starts with the last index
		int nums1Idx = nums1.size() - 1;
		int nums2Idx = nums2.size() - 1;

		//Evaluates the element values from each array, and decides which element is going to be stored in nums1 array
		while (i >= 0 && nums1Idx >= 0) {
		
			//Evaluates the elements from the nums1 and nums2 array
			if (nums2Idx >= 0 && nums1[i] < nums2[nums2Idx]) {
				
				nums1[i] = nums2[nums2Idx];

				--nums2Idx;
				--i;
			}

			//Evaluates the elements from the nums1 array
			else if (nums1Idx >= 0 && nums1[i] > nums1[nums1Idx])  {
				
				int tempNum = nums1[i];

				nums1[i] = nums1[nums1Idx];
				nums1[nums1Idx] = tempNum;

				--nums1Idx;
				--i;
			}
			else if (i == nums1Idx) {
				--nums1Idx;
				--i;
			}
			else {
				--nums1Idx;
			}

		}

	}


	cout << "\n" << "merge() function reached the end!" << "\n\n";
}







/*
void Solution::merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {

	vector<int> nums1tmp;
	nums1tmp.assign(nums1.size(), 0);

	// Ensures the following constraints are met:
	// * nums1.length == m + n
	// * nums2.length == n
	// * nums2.empty == false
	if (nums1.size() == (m + n) && !nums2.empty() && nums2.size() == n) {

		// Copies the elements from nums1 array to the temp array
		for (int i = 0; i < nums1.size(); ++i) {
			nums1tmp[i] = nums1[i];
		}

		int i = 0;
		int nums1Cmp_Idx = 0;
		int nums2Cmp_Idx = 0;

		while(i < nums1.size()) {

			if (nums1tmp[nums1Cmp_Idx] <= nums2[nums2Cmp_Idx] && nums1tmp[nums1Cmp_Idx] != 0) {

				nums1[i] = nums1tmp[nums1Cmp_Idx];
				++nums1Cmp_Idx;
				++i;
			}
			else {
				nums1[i] = nums2[nums2Cmp_Idx];
				++nums2Cmp_Idx;
				++i;
			}
		}


	}
	cout << "\n" << "merge() function reached the end!" << "\n\n";
}


*/